backend/src/main/scala/cromwell/backend/wdl/OutputEvaluator.scala

```diff
@@ -1,71 +1,80 @@
-  
+
   type OutputResult[A] = EitherT[Try, NonEmptyList[String], A]
 
   def evaluateOutputs(jobDescriptor: BackendJobDescriptor,
                       ioFunctions: IoFunctionSet,
                       postMapper: WdlValue => Try[WdlValue] = v => Success(v)): EvaluatedJobOutputs = {
-    val knownValues = jobDescriptor.inputDeclarations map {
-      case (declaration, value) => declaration.name -> value
-    }
+    val knownValues = jobDescriptor.localInputs
     
-    def evaluateOutputExpression(expression: WomExpression): OutputResult[WdlValue] = {
-      EitherT { Try(expression.evaluateValue(knownValues, ioFunctions)).map(_.toEither) }
+    def foldFunction(accumulatedOutputs: Try[ErrorOr[List[(String, WdlValue)]]], output: OutputDefinition) = accumulatedOutputs flatMap { accumulated =>
+      // Extract the valid pairs from the job outputs accumulated so far, and add to it the inputs (outputs can also reference inputs)
+      val allKnownValues = accumulated match {
+        case Valid(outputs) => outputs.toMap[String, WdlValue] ++ knownValues
+        case Invalid(_) => knownValues
+      }
+
+      // Attempt to evaluate the expression using all known values
+      def evaluateOutputExpression: OutputResult[WdlValue] = {
+        EitherT { Try(output.expression.evaluateValue(allKnownValues, ioFunctions)).map(_.toEither) }
+      }
+
+      // Attempt to coerce the wdlValue to the desired output type
+      def coerceOutputValue(wdlValue: WdlValue, coerceTo: WdlType): OutputResult[WdlValue] = {
+        fromEither[Try](
+          // TODO WOM: coerceRawValue should return an ErrorOr
+          coerceTo.coerceRawValue(wdlValue).toEither.leftMap(t => NonEmptyList.one(t.getMessage))
+        )
+      }
+
+      /*
+        * Go through evaluation, coercion and post processing.
+        * Transform the result to a validated Try[ErrorOr[(String, WdlValue)]] with toValidated
+        * If we have a valid pair, add it to the previously accumulated outputs, otherwise combine the Nels of errors
+       */
+      (for {
+        evaluated <- evaluateOutputExpression
+        coerced <- coerceOutputValue(evaluated, output.womType)
+        postProcessed <- fromEither[Try](postMapper(coerced).toErrorOr.toEither)
+        pair = output.name -> postProcessed
+      } yield pair).toValidated map { evaluatedOutput: ErrorOr[(String, WdlValue)] =>
+        (accumulated, evaluatedOutput) mapN { _ :+ _ }
+      }
     }
     
-    def coerceOutputValue(wdlValue: WdlValue, coerceTo: WdlType): OutputResult[WdlValue] = {
-      fromEither[Try](
-        coerceTo.coerceRawValue(wdlValue).toEither
-        .leftMap(t => NonEmptyList.one(t.getMessage))
+    val emptyValue = Success(List.empty[(String, WdlValue)].validNel): Try[ErrorOr[List[(String, WdlValue)]]]
+
+    // Fold over the outputs to evaluate them in order, map the result to an EvaluatedJobOutputs
+    jobDescriptor.call.callable.outputs.foldLeft(emptyValue)(foldFunction) match {
```

Comment:
ToL so much IntelliJ red here that seems bogus to me 😢

Reply:
oh I don't have any red here : /